xtask/src/release/changelog.rs

```diff
@@ -0,0 +1,159 @@
+use std::fmt::Write;
+use std::{env, iter};
+
+use anyhow::{bail, Result};
+use xshell::cmd;
+
+pub(crate) fn get_changelog(
+    changelog_n: usize,
+    commit: &str,
+    prev_tag: &str,
+    today: &str,
+) -> Result<String> {
+    let git_log = cmd!("git log {prev_tag}..HEAD --merges --reverse").read()?;
+    let mut features = String::new();
+    let mut fixes = String::new();
+    let mut internal = String::new();
+    let mut others = String::new();
+    for line in git_log.lines() {
+        let line = line.trim_start();
+        if let Some(p) = line.find(':') {
+            let pr = &line[..p];
+            if let Ok(pr_num) = pr.parse::<u32>() {
+                let accept = "Accept: application/vnd.github.v3+json";
+                let token = match env::var("GITHUB_TOKEN") {
+                    Ok(token) => token,
+                    Err(_) => bail!("Please obtain a personal access token from https://github.com/settings/tokens and set the `GITHUB_TOKEN` environment variable."),
+                };
+                let authorization = format!("Authorization: token {}", token);
+                let pr_url = "https://api.github.com/repos/rust-analyzer/rust-analyzer/issues";
+
+                // we don't use an HTTPS client or JSON parser to keep the build times low
+                let pr_json =
+                    cmd!("curl -s -H {accept} -H {authorization} {pr_url}/{pr}").read()?;
```

Comment:
Hm, why do we need to curl the API here? Isn't this info available locally? I think bors copy-pastes PR description / title into the merge commint?

Reply:
To get the title and description in case they were edited after merging. I did that for some of them to reduce the notification spam (but arguably that might not help too much because the email client might split the thread when the subject changes).

Reply:
I wouldn't expect email clients to split threads on the subject but on the `In-Reply-To` header.

Reply:
I haven't paid attention, but in Gmail it happens when someone replies via email. It might also happen when you edit the title.

At least it works if you keep an empty inbox.